trpcultivate_phenotypes/src/Service/TripalCultivatePhenotypesGenusProjectService.php

```diff
@@ -0,0 +1,154 @@
+   * Configuration terms.genus.
+   */
+  private $sysvar_genus;
+
+  /**
+   * Configuration genus.ontology.
+   */
+  private $sysvar_genusontology;
+
+  /**
+   * Chado database and Tripal logger.
+   */
+  protected $chado;
+  protected $logger;
+
+  /**
+   * Constructor.
+   */
+  public function __construct(ConfigFactoryInterface $config_factory, ChadoConnection $chado, TripalLogger $logger) {
+    // Module Configuration variables.
+    $module_settings = 'trpcultivate_phenotypes.settings';
+    $config = $config_factory->getEditable($module_settings);
+
+    // Configuration terms.genus.
+    $this->sysvar_genus = $config->get('trpcultivate.phenotypes.ontology.terms.genus');
+    // Configuration genus.ontology.
+    $this->sysvar_genusontology = $config->get('trpcultivate.phenotypes.ontology.cvdbon');
+
+    // Chado database.
+    $this->chado = $chado;
+
+    // Tripal Logger service.
+    $this->logger = $logger;
+  }
+  
+  /**
+   * Assign a genus to an experiment/project.
+   * 
+   * @param int $project
+   *   Project (project id number) the parameter $genus will be assigned to.
+   * @param string $genus
+   *   Genus name/title.
+   * @param boolean $replace
+   *   True to replace existing genus of a project with a different genus.
+   *   Default to False.
+   *
+   * @return boolean
+   *   True, genus was set successfully or false on error/fail.
+   */
+  function setGenusToProject($project, $genus, $replace = FALSE) {
+    $error = 0;
+
+    if ($project > 0 && !empty($genus)) {
+      $result = $this->chado->query("
+        SELECT projectprop_id AS id FROM {1:projectprop} 
+        WHERE project_id = :project_id AND type_id = :type_id LIMIT 1
+      ", [':project_id' => $project, ':type_id' => $this->sysvar_genus]);
+      
+      $projectprop_id = $result->fetchField();
+
+      if ($projectprop_id > 0) {
+        // Has a genus.
+        if ($replace) {
+          // And wishes to replace with another genus.
+          $this->chado->query("
+            UPDATE {1:projectprop} SET value = :new_genus WHERE projectprop_id = :id
+          ", [':new_genus' => $genus, ':id' => $projectprop_id]);
+        }
+
+        // Do nothing if maintain the same genus.
+      } 
+      else {
+        // Not set yet, no record in projectprop.
+        // Create a relationship regardless to replace or not.
+        $sql = "INSERT INTO {1:projectprop} (project_id, type_id, value) VALUES ('%s', '%s', '%s')";
+        $query = sprintf($sql, $project, $this->sysvar_genus, $genus);
+        $this->chado->query($query);
+      }
+    }
+    else {
+      $error = 1;
+      $this->logger->error('Error. Could not replace genus. Invalid project or genus');
+    }
+
+    return ($error) ? FALSE : TRUE;
+  }
+
+  /**
+   * Get genus of an experiment/project.
+   * 
+   * @param int $project
+   *   Project (project_id number) to search.
+   *
+   * @return array
+   *   Key is genus/organism id number and value is the genus name/title.    
```

Comment:
This should only return the genus. That is because there may be more than one organism with the same genus. Here you are arbitrarily taking the first organism entry that postgresql gives you 😬

Reply:
Additionally, for performance reasons I think we select directly from project prop without confirming the genus exists or is configured.

Reply:
The query already restricts genus based on the genus set in projectprop. I have also updated the query to only consider genus that are configured.  Similary, I have also added a check in the set genus method to also check the genus (that it is configured) before pairing it to a project. Hopefully both measures will improve accuracy of returned genus and prevent inactive genus from being used.